### PR TITLE
[Stable] Add support for role variables

### PR DESCRIPTION
**Pull request by kasper-piskorski**

### CODE REVIEW

**Comment by kasper-piskorski** [comment]
!rtg- grakn-dist failure

**Comment by kasper-piskorski** [comment]
!rtg

**Comment by kasper-piskorski** [comment]
!rtg - ai.grakn.factory.TxFactoryJanusTest.ai.grakn.factory.TxFactoryJanusTest

File: grakn-core/src/main/java/ai/grakn/graql/admin/Answer.java [hunk 1]
```diff
@@ -147,6 +147,13 @@
     Stream<Answer> permute(Set<Unifier> unifierSet);
 
     /**
+     * @param vars variables for which role hierarchy should be expanded
```
Comment (fppt) [inline grakn-core/src/main/java/ai/grakn/graql/admin/Answer.java]: use `{@link Role}` when referring to roles please

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationAtom.java [hunk 2]
```diff
@@ -283,6 +293,18 @@ public int computePriority(Set<Var> subbedVars) {
...
 
+    public Stream<IdPredicate> getRolePredicates(){
+        return getRelationPlayers().stream()
+                .map(RelationPlayer::getRole)
+                .flatMap(CommonUtil::optionalToStream)
+                .filter(var -> var.var().isUserDefinedName())
+                .filter(vp -> vp.getTypeLabel().isPresent())
+                .map(vp -> {
+                    Label label = vp.getTypeLabel().orElse(null);
+                    return new IdPredicate(vp.var(), tx().getRole(label.getValue()), getParentQuery());
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationAtom.java]: According to the above line: `label.getValue()` is risking a null pointer exception. Might be better to throw a more meaningful error or revisit the logic and see if the null can be avoided.

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationAtom.java [hunk 3]
```diff
@@ -489,12 +517,27 @@ public Atom inferTypes() {
...
-        return unmappedVars;
+    private Set<Pair<Var, Var>> getNonSpecificRelationPlayers() {
+        return getRelationPlayers().stream()
+                .filter(rp -> rp.getRole().isPresent())
+                .filter(rp -> {
+                    VarPatternAdmin rolePattern = rp.getRole().orElse(null);
+                    if (rolePattern == null) return true;
+
+                    Label roleLabel = rolePattern.getTypeLabel().orElse(null);
+                    return roleLabel == null || Schema.MetaSchema.isMetaLabel(roleLabel);
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationAtom.java]: This null check `roleLabel == null` is not needed. `Schema.MetaSchema.isMetaLabel(roleLabel)` can take in a null and fail as expected.

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationAtom.java [hunk 4]
```diff
@@ -773,14 +813,59 @@ public Unifier getUnifier(Atom pAtom) {
...
+
+                    });
         }
         return unifier.removeTrivialMappings();
     }
 
-    @Override
-    public Atom rewriteToUserDefined(){
+    /**
+     * if any role variable of the parent is user defined rewrite ALL role variables to user defined (otherwise unification is problematic)
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationAtom.java]: `{@link Role}`

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/query/ReasonerAtomicQuery.java [hunk 5]
```diff
@@ -321,14 +323,16 @@ public Answer lookupAnswer(QueryCache<ReasonerAtomicQuery> cache, Answer sub) {
             while(ruleIterator.hasNext()) {
                 RuleTuple ruleContext = ruleIterator.next();
                 InferenceRule rule = ruleContext.getRule();
-                Unifier u = ruleContext.getRuleUnifier();
-                Unifier pu = ruleContext.getPermutationUnifier();
+                Unifier unifier = ruleContext.getRuleUnifier();
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/query/ReasonerAtomicQuery.java]: 👍

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/state/RoleExpansionState.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import ai.grakn.graql.admin.Answer;
+import ai.grakn.graql.admin.Unifier;
+import ai.grakn.graql.internal.reasoner.query.ReasonerAtomicQuery;
+import java.util.Iterator;
+import java.util.Set;
+
+/**
+ *
+ * <p>
+ * Query state produced by {@link AtomicState} when an atomic query {@link ReasonerAtomicQuery}contains role variables which may require role hierarchy expansion.
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/state/RoleExpansionState.java]: `{@link Role}`

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/utils/ReasonerUtils.java [hunk 7]
```diff
@@ -202,6 +207,20 @@ public static IdPredicate getIdPredicate(Var typeVariable, VarPatternAdmin typeV
     }
 
     /**
+     * @param concept which hierarchy should be considered
+     * @return set of schema concepts: provided concept and all its supers including meta
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/utils/ReasonerUtils.java]: `set of schema concepts` -> `set of {@link SchemaConcept}s`

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationAtom.java (hunk 2 above)
Comment (kasper-piskorski) [reply grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationAtom.java]: There is a IsPresent above.  The reason I didn't collapse the optional in the stream is that i need both the label and the vp.

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationAtom.java (hunk 3 above)
Comment (kasper-piskorski) [reply grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationAtom.java]: if `roleLabel` is null I need it to return `true`. If I pass null to `isMetaLabel` it returns `false`.

**Pull request merged**